### ISSUE TITLE
[portfolio allocation] Model doesn't seem to really make use of cov_list

### ISSUE THREAD

**Issue by ihopethiswillfi** [AI4Finance-Foundation/FinRL#340]
Hi, I'm looking at `FinRL_PortfolioAllocation_NeurIPS_2020`.

Just to improve my understanding a bit, in the StockPortfolioEnv I decided to replace all mentions of `cov_list` with `return_list` (I had to do 3 replacements in total).  So instead of feeding the cov values to the model, we are giving it the historic returns instead.

```
#self.covs = self.data["cov_list"].values[0]     # old
self.covs = self.data["return_list"].values[0]   # new
```

In order to get this working, I naturally had to update the dimensions of `self.state` cause cov_list dimensions are (28,28) and return_list dimensions are (252,28):
```
self.state =  np.append(np.array(self.covs), [self.data[tech].values.tolist() for tech in self.tech_indicator_list ], axis=0)
```

Doing the above barely had any impact on the skill of the model, so I decided to try another experiment.

I inserted the below into `step()`:
```
# experiment: we leak the return of 1 day into the future
if self.day < self.df.index.max():
    self.covs = self.covs.shift(-1)
    covsnextday = self.df.loc[self.day+1, :]["return_list"].values[0]
    self.covs.iloc[-1] = covsnextday.iloc[-1]
```
The obvious expectation here is that we now get to see superb performance, because we are leaking the returns of the next step.
But this didn't really happen.  It actually had no impact at all!

I have tested this with A2C:
- before any changes:  2.155 sharpe on trading (validation) set
- after changing cov to return_list: 2.120 sharpe on trading (validation) set
- after changing cov to return_list (with leak): 2.049 sharpe on trading (validation) set

My conclusion is that it looks like the model isn't really making proper use of `self.covs` , which I don't really understand because self.covs is part of self.state.

Perhaps I'm overlooking something, or my entire assesment is wrong?  Anything else you'd like me to test to further confirm this?

**Comment by rayrui312** [comment]
Thanks for reporting this issue. What hyperparameters are you using? I think it may be because the hyperparameters are not tuned well. And could you test it using PPO algorithm?

**Comment by ihopethiswillfi** [comment]
This is with the default parameters in the mentioned notebook.  
I don't think I made any other change to that notebook other than those I mentioned above, so it's very easy to reproduce.
```
A2C_PARAMS = {"n_steps": 5, "ent_coef": 0.005, "learning_rate": 0.0002}
model_a2c = agent.get_model(model_name="a2c",model_kwargs = A2C_PARAMS)
```

I have now tested the results with the other algorithms too, all with the default parameters from the notebook:
```
df_daily_return, df_actions = DRLAgent.DRL_prediction(model=trained_a2c,
                        environment = e_trade_gym)
df_daily_return, df_actions = DRLAgent.DRL_prediction(model=trained_ppo,
                        environment = e_trade_gym)
df_daily_return, df_actions = DRLAgent.DRL_prediction(model=trained_ddpg,
                        environment = e_trade_gym)
df_daily_return, df_actions = DRLAgent.DRL_prediction(model=trained_sac,
                        environment = e_trade_gym)
df_daily_return, df_actions = DRLAgent.DRL_prediction(model=trained_td3,
                        environment = e_trade_gym)
```

Results:
```
=================================
begin_total_asset:1000000
end_total_asset:1427516.2877314668
Sharpe:  1.9648665204307265
=================================
hit end!
=================================
begin_total_asset:1000000
end_total_asset:1427569.807797934
Sharpe:  2.0142868804423797
=================================
hit end!
=================================
begin_total_asset:1000000
end_total_asset:1409876.3863395497
Sharpe:  2.0133478227135164
=================================
hit end!
=================================
begin_total_asset:1000000
end_total_asset:1433626.7209164905
Sharpe:  2.0341284475446146
=================================
hit end!
=================================
show more (open the raw output data in a text editor) ...

begin_total_asset:1000000
end_total_asset:1389737.2130779005
Sharpe:  1.923941754990748
=================================
hit end!
```

Above results slightly change every time I re-run the cell.

None of the models seem to be exploiting the leak.

**Comment by rayrui312** [comment]
I think it is because it doesn't do feature scaling. The magnitude of these numbers is too tiny compared to other numbers in the states, such as the prices.

**Comment by ihopethiswillfi** [comment]
Good point!  I had a look at this.
After comparing the min, mean, max of self.covs and self.tech_indicator_list, I decided to multiply covs by 1000 when doing the assignment to self.state:
```
self.state =  np.append(np.array(self.covs)*1000, [self.data[tech].values.tolist() for tech in self.tech_indicator_list ], axis=0)
```
I did this in __init__ , step and reset.

After doing another run, the validation numbers didn't change much.  A2C is slightly worse.  PPO is slightly improved.  Totally in line with doing a run with another seed.
